### ISSUE TITLE
A/F: `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' after f12a556

### ISSUE THREAD

**Issue by dyung** [llvm/llvm-project#64040]
One of our internal tests started to hit an assertion failure, and I was able to bisect it back to commit f12a556. Consider the following code:
```c++
typedef int a;
typedef float __attribute__((ext_vector_type(4))) b;
typedef a __attribute__((ext_vector_type(4))) c;
typedef int __attribute__((ext_vector_type(4))) d;
typedef float __attribute__((ext_vector_type(8))) e;
#define f() 0
#define g(h, i) (h < i)
void j(c);
b k;
void l() {
  d m{4};
  for (char n = 0; n < 4; ++n) {
    e o = f();
    b p = o.r;
    d q g(k, p);
    m -= q;
  }
  j(m);
}
```
If this is compiled with optimizations enabled, the compiler hits an assertion failure:
```
$ ~/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang -O2 -c repro2.cpp
clang: /home/dyung/src/upstream/llvm_clean_git/llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From*) [with To = llvm::IntegerType; From = const llvm::Type]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang -O2 -c repro2.cpp
1.	<eof> parser at end of file
2.	Optimizer
 #0 0x00005618e079b9ef llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3bf29ef)
 #1 0x00005618e079975c llvm::sys::CleanupOnSignal(unsigned long) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3bf075c)
 #2 0x00005618e06e2038 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007f3a54432420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #4 0x00007f3a53eff00b raise /build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007f3a53ede859 abort /build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:81:7
 #6 0x00007f3a53ede729 get_sysdep_segment_value /build/glibc-SzIz7B/glibc-2.31/intl/loadmsgcat.c:509:8
 #7 0x00007f3a53ede729 _nl_load_domain /build/glibc-SzIz7B/glibc-2.31/intl/loadmsgcat.c:970:34
 #8 0x00007f3a53eeffd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #9 0x00005618e02b9a8b llvm::InstCombinerImpl::foldBinOpOfSelectAndCastOfSelectCondition(llvm::BinaryOperator&)::'lambda0'(bool, llvm::Value*)::operator()(bool, llvm::Value*) const InstructionCombining.cpp:0:0
#10 0x00005618e02bf866 llvm::InstCombinerImpl::foldBinOpOfSelectAndCastOfSelectCondition(llvm::BinaryOperator&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3716866)
#11 0x00005618e02eeb77 llvm::InstCombinerImpl::visitAdd(llvm::BinaryOperator&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3745b77)
#12 0x00005618e02cf920 llvm::InstCombinerImpl::run() (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3726920)
#13 0x00005618e02d233b combineInstructionsOverFunction(llvm::Function&, llvm::InstructionWorklist&, llvm::AAResults*, llvm::AssumptionCache&, llvm::TargetLibraryInfo&, llvm::TargetTransformInfo&, llvm::DominatorTree&, llvm::OptimizationRemarkEmitter&, llvm::BlockFrequencyInfo*, llvm::ProfileSummaryInfo*, unsigned int, llvm::LoopInfo*) InstructionCombining.cpp:0:0
#14 0x00005618e02d31ee llvm::InstCombinePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x372a1ee)
#15 0x00005618e0a05c56 llvm::detail::PassModel<llvm::Function, llvm::InstCombinePass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3e5cc56)
#16 0x00005618ddffe73d llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x145573d)
#17 0x00005618df6c5931 llvm::CGSCCToFunctionPassAdaptor::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x2b1c931)
#18 0x00005618ddfe2536 llvm::detail::PassModel<llvm::LazyCallGraph::SCC, llvm::CGSCCToFunctionPassAdaptor, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x1439536)
#19 0x00005618df6bdb2f llvm::PassManager<llvm::LazyCallGraph::SCC, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x2b14b2f)
#20 0x00005618e1aa7e86 llvm::detail::PassModel<llvm::LazyCallGraph::SCC, llvm::PassManager<llvm::LazyCallGraph::SCC, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x4efee86)
#21 0x00005618df6c1f5d llvm::DevirtSCCRepeatedPass::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x2b18f5d)
#22 0x00005618e1aa7ed6 llvm::detail::PassModel<llvm::LazyCallGraph::SCC, llvm::DevirtSCCRepeatedPass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x4efeed6)
#23 0x00005618df6c002d llvm::ModuleToPostOrderCGSCCPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x2b1702d)
#24 0x00005618e1c6fea7 llvm::ModuleInlinerWrapperPass::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x50c6ea7)
#25 0x00005618e1aa7616 llvm::detail::PassModel<llvm::Module, llvm::ModuleInlinerWrapperPass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x4efe616)
#26 0x00005618e0142a7d llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3599a7d)
#27 0x00005618e0a16f57 (anonymous namespace)::EmitAssemblyHelper::RunOptimizationPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>&) BackendUtil.cpp:0:0
#28 0x00005618e0a19f71 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3e70f71)
#29 0x00005618e1a4675d clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x4e9d75d)
#30 0x00005618e2bf0fe9 clang::ParseAST(clang::Sema&, bool, bool) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x6047fe9)
#31 0x00005618e1a44f28 clang::CodeGenAction::ExecuteAction() (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x4e9bf28)
#32 0x00005618e1285b39 clang::FrontendAction::Execute() (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x46dcb39)
#33 0x00005618e120a706 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x4661706)
#34 0x00005618e136a316 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x47c1316)
#35 0x00005618ddc35a44 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x108ca44)
#36 0x00005618ddc2e08a ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#37 0x00005618e1062e8d void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#38 0x00005618e06e2540 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x3b39540)
#39 0x00005618e10634af clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#40 0x00005618e1029afc clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x4480afc)
#41 0x00005618e102a58d clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x448158d)
#42 0x00005618e10326ed clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x44896ed)
#43 0x00005618ddc33caa clang_main(int, char**, llvm::ToolContext const&) (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x108acaa)
#44 0x00005618ddb49225 main (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0xfa0225)
#45 0x00007f3a53ee0083 __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:342:3
#46 0x00005618ddc2db4e _start (/home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin/clang+0x1084b4e)
clang: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 17.0.0 (https://github.com/llvm/llvm-project.git f12a5561b2cbfae384c9a31293938ee2acea79fd)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/dyung/src/upstream/f12a5561b2cbfae384c9a31293938ee2acea79fd-linux/bin
```
@antoniofrighetto, can you take a look since this was your change which caused this?

**Comment by Benjins** [comment]
Looks like it's hitting the same issue as https://github.com/llvm/llvm-project/issues/64036

**Comment by antoniofrighetto** [comment]
Thanks for reporting to you both, should be fixed by ac3f6899f45cfbaf53fa503e4100e92a731b2e7b. A remind to myself to test deeper SIMD code generation.
